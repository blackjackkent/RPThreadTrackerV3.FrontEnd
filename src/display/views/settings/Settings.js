import React from 'react';
import { Row, Col } from 'reactstrap';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import ChangePasswordPane from './components/ChangePasswordPane';
import UpdateAccountInfoPane from './components/UpdateAccountInfoPane';
import DeleteAccountPane from './components/DeleteAccountPane';
import * as actions from '../../../infrastructure/actions';
import TabContent from '../../shared/styled/TabContent';
import StaticTabNav from '../../shared/static/StaticTabNav';
import StaticDropdownNav from '../../shared/static/StaticDropdownNav';
import tabs from '../../../infrastructure/constants/tabs';

const propTypes = {
	user: PropTypes.shape({}).isRequired,
	submitUserChangePassword: PropTypes.func.isRequired,
	submitUserAccountInfo: PropTypes.func.isRequired,
	openDeleteAccountConfirmationModal: PropTypes.func.isRequired,
	match: PropTypes.shape({
		url: PropTypes.string,
		params: PropTypes.shape({
			tabId: PropTypes.string
		})
	}).isRequired
};

function mapStateToProps(state) {
	const { user } = state;
	return {
		user
	};
}

const Settings = (props) => {
	const {
		user,
		submitUserChangePassword,
		submitUserAccountInfo,
		openDeleteAccountConfirmationModal,
		match
	} = props;
	const options = Object.values(tabs.SETTINGS);
	const onDeleteAccountClicked = (e) => {
		e.preventDefault();
		openDeleteAccountConfirmationModal();
	};
	return (
		<div className="animated fadeIn static-container settings-container">
			<Row>
				<Col className="d-lg-none text-center">
					<StaticDropdownNav
						data-spec="settings-static-dropdown-nav"
						activeTab={match.url}
						options={options}
					/>
				</Col>
			</Row>
			<Row>
				<Col className="d-none d-lg-block" md={3}>
					<StaticTabNav data-spec="settings-static-tab-nav" options={options} />
				</Col>
				<Col xs="12" lg="9">
					<TabContent activeTab={match.params.tabId}>
						<ChangePasswordPane submitChangePasswordForm={submitUserChangePassword} />
						<UpdateAccountInfoPane
							submitAccountInfoForm={submitUserAccountInfo}
							user={user}
						/>
						<DeleteAccountPane
							data-spec="delete-account-pane"
							onDeleteAccountClicked={onDeleteAccountClicked}
						/>
					</TabContent>
				</Col>
			</Row>
		</div>
	);
};

Settings.propTypes = propTypes;
export default connect(mapStateToProps, {
	submitUserChangePassword: actions.submitUserChangePassword,
	submitUserAccountInfo: actions.submitUserAccountInfo,
	openDeleteAccountConfirmationModal: actions.openDeleteAccountConfirmationModal
})(Settings);
