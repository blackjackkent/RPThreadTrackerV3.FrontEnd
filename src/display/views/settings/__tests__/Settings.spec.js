// #region imports
import React from 'react';
import { getSpecWrapper, shallowWithState } from '~/testhelpers/helpers.unit';
import Settings from '../Settings';
// #endregion imports

// #region mocks
jest.mock('../components/ChangePasswordPane', () => 'ChangePasswordPane');
jest.mock('../components/UpdateAccountInfoPane', () => 'UpdateAccountInfoPane');
jest.mock('../../../../infrastructure/actions', () => ({}));
jest.mock('../../../../infrastructure/constants/tabs', () => ({
	SETTINGS: {
		TAB1: {
			href: '/settings/tab1',
			name: 'Tab 1'
		},
		TAB2: {
			href: '/settings/tab2',
			name: 'Tab 2'
		}
	}
}));
jest.mock('../../../shared/static/StaticTabNav', () => 'StaticTabNav');
jest.mock('../../../shared/static/StaticDropdownNav', () => 'StaticDropdownNav');
// #endregion mocks

const createTestProps = (propOverrides) => ({
	openDeleteAccountConfirmationModal: jest.fn(),
	submitUserChangePassword: jest.fn(),
	submitUserAccountInfo: jest.fn(),
	match: {
		url: '/settings/tab1',
		params: {
			tabId: 'tab1'
		}
	},
	...propOverrides
});

const createTestState = (stateOverrides) => ({
	user: {
		id: '12345'
	},
	...stateOverrides
});

describe('rendering', () => {
	describe('snapshots', () => {
		it('should render valid snapshot', () => {
			const props = createTestProps();
			const state = createTestState();
			const jsx = <Settings {...props} />;
			const element = shallowWithState(jsx, state).dive();
			expect(element).toMatchSnapshot();
		});
	});
	describe('options', () => {
		it('should populate options from imported tabs on dropdown nav', () => {
			const props = createTestProps();
			const state = createTestState();
			const jsx = <Settings {...props} />;
			const element = shallowWithState(jsx, state).dive();
			const form = getSpecWrapper(element, 'settings-static-dropdown-nav');
			const { options } = form.props();
			expect(options).toHaveLength(2);
		});
		it('should populate options from imported tabs on tab nav', () => {
			const props = createTestProps();
			const state = createTestState();
			const jsx = <Settings {...props} />;
			const element = shallowWithState(jsx, state).dive();
			const form = getSpecWrapper(element, 'settings-static-tab-nav');
			const { options } = form.props();
			expect(options).toHaveLength(2);
		});
	});
	describe('onDeleteAccountClicked', () => {
		it('should trigger modal open action', () => {
			const openDeleteAccountConfirmationModal = jest.fn();
			const props = createTestProps({
				openDeleteAccountConfirmationModal
			});
			const state = createTestState();
			const jsx = <Settings {...props} />;
			const element = shallowWithState(jsx, state).dive();
			const deletePane = getSpecWrapper(element, 'delete-account-pane');
			deletePane.props().onDeleteAccountClicked({ preventDefault: jest.fn() });
			expect(openDeleteAccountConfirmationModal).toHaveBeenCalledTimes(1);
		});
	});
});
