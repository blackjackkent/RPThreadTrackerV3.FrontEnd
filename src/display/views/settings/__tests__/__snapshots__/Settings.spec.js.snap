// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering snapshots should render valid snapshot 1`] = `
<div
  className="animated fadeIn static-container settings-container"
>
  <Row
    tag="div"
  >
    <Col
      className="d-lg-none text-center"
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <StaticDropdownNav
        activeTab="/settings/tab1"
        data-spec="settings-static-dropdown-nav"
        options={
          Array [
            Object {
              "href": "/settings/tab1",
              "name": "Tab 1",
            },
            Object {
              "href": "/settings/tab2",
              "name": "Tab 2",
            },
          ]
        }
      />
    </Col>
  </Row>
  <Row
    tag="div"
  >
    <Col
      className="d-none d-lg-block"
      md={3}
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <StaticTabNav
        data-spec="settings-static-tab-nav"
        options={
          Array [
            Object {
              "href": "/settings/tab1",
              "name": "Tab 1",
            },
            Object {
              "href": "/settings/tab2",
              "name": "Tab 2",
            },
          ]
        }
      />
    </Col>
    <Col
      lg="9"
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
      xs="12"
    >
      <Styled(TabContent)
        activeTab="tab1"
      >
        <ChangePasswordPane
          submitChangePasswordForm={[MockFunction]}
        />
        <UpdateAccountInfoPane
          submitAccountInfoForm={[MockFunction]}
          user={
            Object {
              "id": "12345",
            }
          }
        />
        <DeleteAccountPane
          data-spec="delete-account-pane"
          onDeleteAccountClicked={[Function]}
        />
      </Styled(TabContent)>
    </Col>
  </Row>
</div>
`;
